README.md
# GG EZ Video Player

Simple video player with minimal setup, intuitive API and straightforward features.

![Player](./images/player.png)

## Features:

-   Easy setup
-   Customizable controls (or use our slick defaults!)
-   Lightweight
-   Programmatic access to video playback APIs
-   Browser (UMD), CommonJS and ES module versions
-   VAST / VPAID Support for a single Linear Creative. Parsing is provided by [dailymotion/vast-client-js](https://github.com/dailymotion/vast-client-js). _Note_ that at the time, only one Ad is loaded, and only Linear Creatives are played and tracked.

[See the demo page](https://gumgum.github.io/gg-ez-vp/demo)

## Table of contents

-   [Features](#features)
-   [Table of contents](#table-of-contents)
-   [Installation](#installation)
    -   [Browser](#browser)
    -   [CommonJS](#commonjs)
    -   [ES Modules](#es-modules)
-   [Usage](#usage)
-   [Customization](#customization)
    -   [Options](#options)
    -   [Controls](#controls)
        -   [Control configuration examples](#control-configuration-examples)
    -   [Styles](#styles)
-   [Public methods](#public-methods)
-   [Events](#events)
-   [VAST](#vast)
-   [VPAID](#vpaid)
    -   [Event mapping](#event-mapping)
-   [Development](#development)

## Installation

#### Browser

```
<!-- latest stable version -->
<script src="https://c.gumgum.com/vp/latest/gg-ez-vp.js">

<!-- beta version | synced with develop branch  -->
<script src="https://c.gumgum.com/vp/beta/gg-ez-vp.js">

<!-- or specific release, replace "1.x.x" with the release you want to use -->
<script src="https://c.gumgum.com/vp/1.x.x/gg-ez-vp.js">
```

**Note:** One of [GumGum's solutions, is contextual advertising](https://gumgum.com/advertising), as such, our CDN is blocked by ad blockers; if the scripts are not loading, disable the ad blocker or use an external CDN (TBD #21).

#### CommonJS

TODO: Installation

```
const GgEzVp = require('gg-ez-vp');
```

#### ES Modules

TODO: Installation

```
import GgEzVp from 'gg-ez-vp';
```

## Usage

GgEzVp works by attaching itself to a unique container, ideally an empty `<div>` with a unique id:

```
<body>
    <div>
        <div class="randomContent">
            Some content
        </div>
        <div id="myVideo" class="myVideoStyles"></div>
    </div>
</body>
```

Then, to create a player instance:

```
// js
const minimalConfiguration = {
    container: 'myVideo',
    src: 'myvideo.mp4'
};

const ggEzVpInstance = new GgEzVp(minimalConfiguration);
```

This minimal configuration will create a unique video player instance using the defaults set by GgEzVp.

**note**: `src` and `container` are the only required options in the configuration object.

## Customization

### Options:

| key         | defaultValue | required | description                                                                                                                                                                           |
| ----------- | ------------ | -------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| container   | null         | true     | either a node's id or the DOM node itself where the player will be loaded to                                                                                                          |
| src         | null         | true     | string or array of strings used to retrieve video and/or VAST                                                                                                                         |
| width       | null         | false    | by default the player will take available space, unless width is specified                                                                                                            |
| height      | null         | false    | by default the player will take available space, unless height is specified                                                                                                           |
| controls    | object       | false    | specific controls configurations, see Controls section below                                                                                                                          |
| autoplay    | false        | false    | whether to play the video automatically or not                                                                                                                                        |
| volume      | 1            | false    | initial volume for playback must be between 0.0 and 1                                                                                                                                 |
| muted       | true         | false    | video will be muted by default                                                                                                                                                        |
| playsinline | true         | false    | [A Boolean attribute indicating that the video is to be played within the element's playback area.](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video#attr-playsinline) |
| poster      | null         | false    | source for an image to be used as video poster                                                                                                                                        |
| preload     | 'auto'       | false    | standard HTML values for preload (none                                                                                                                                                | metadata | auto) |
| loop        | false        | false    | whether to loop video or not                                                                                                                                                          |
| isVAST      | false        | false    | enables support for a single VAST / VPAID TAG to be parsed and used as source, will force displaying the ad controls                                                                  |
| adControls  | false        | false    | forces displaying the ad controls instead of the regular contols                                                                                                                      |

### Controls

Besides the programmatic interaction with the player, there are also a number of controls displayed by default, they can be toggled by passing a `controls` property in the initial configuration.
See the Styles section for information on how to customize the visuals.

![Player](./images/player.png)

| controlName | defaultValue | description                                                                                                                      |
| ----------- | ------------ | -------------------------------------------------------------------------------------------------------------------------------- |
| timestamp   | true         | display the current time of the video in the format [m:ss]                                                                       |
| volume      | true         | display the volume controls (mute toggle and range input)                                                                        |
| volumeRange | true         | the volume range input can be hidden if only the volume button is needed                                                         |
| progress    | true         | display the playback progress bar, clicking on the bar will set the video currentTime to the selected percentage (except on Ads) |
| play        | true         | show the play/pause button                                                                                                       |
| expand      | true         | show the expand to fullscreen button                                                                                             |
| timestampAd | false        | **used only for Ads**, will display the remaining time in the format [Ad ss]                                                     |
| skip        | false        | **used only for Ads**, will display a skip button                                                                                |

Additionally, passing the `adControls: true` option, will display a different set of controls for Ads:

![Ad Controls](./images/ad_controls.png)

| controlName | defaultValue | description                                                                                            |
| ----------- | ------------ | ------------------------------------------------------------------------------------------------------ |
| timestampAd | true         | display the remainingTime time of the ad in the format [Ad ss].                                        |
| skip        | true         | display a skip button, pressing it will emit the 'skip' event                                          |
| progress    | true         | the progress bar will display, but it won't show the scrub, nor it be possible to set the time with it |
| play        | true         | show the play/pause button                                                                             |
| expand      | true         | show the expand to fullscreen button                                                                   |

**Note:** the `timestamp` control will be disabled when `adControls` is true and instead `timestampAd` will be displayed.

#### Control configuration examples

```
// Controls off
const controlsOffConfig = {
    container: 'myVideo',
    src: 'myvideo.mp4',
    controls: false
};

// timestamp and volume off
const customControls = {
    container: 'myVideo',
    src: 'myvideo.mp4',
    controls: {
        timestamp: false,
        volume: false
    }
};

// ad timestamp, skip and expand off
const customControls = {
    container: 'myVideo',
    src: 'myvideo.mp4',
    adControls: true, // will enable ad controls
    controls: {
        timestampAd: false,
        expand: false,
        skip: false
    }
};
```

### Styles

To use the default styles, just include the `styles.css` file before loading the player's javascript:

```
<!-- latest version of the styles -->
<link src="https://c.gumgum.com/vp/latest/gg-ez-vp.css">

```

The default styles that can be overridden by any custom styles you may need:

```
.gg-ez-vp--play {
    background: url('/images/my-custom-play-btn.png');
    width: 40px;
    height: 40px;
}

.gg-ez-vp--progress-filled {
    background: red;
}
```

**It is encouraged to override the styles rather than completely writing your own.**

Use these selectors to override the styles:

| selector                         | component                                                                 |
| -------------------------------- | ------------------------------------------------------------------------- |
| .gg-ez-vp                        | player container                                                          |
| .gg-ez-vp--viewer                | video element                                                             |
| .gg-ez-vp--controls              | control bar                                                               |
| .gg-ez-vp--timestamp             | current video time                                                        |
| .gg-ez-vp--timestamp-break       | bar dividing currentime / duration                                        |
| .gg-ez-vp--timestamp-ad          | ad remaining time                                                         |
| .gg-ez-vp--volume                | volume controls container                                                 |
| .gg-ez-vp--volume-control        | volume slider container                                                   |
| .gg-ez-vp--volume-control-slider | volume slider                                                             |
| .gg-ez-vp--input-range           | volume slider input styles                                                |
| .gg-ez-vp--progress              | progress container                                                        |
| .gg-ez-vp--progress-bar          | progress track                                                            |
| .gg-ez-vp--progress-filled       | progress bar fill                                                         |
| .gg-ez-vp--progress-filled:after | progress bar scrub                                                        |
| .gg-ez-vp--button-icon           | play and volume buttons                                                   |
| .gg-ez-vp--button-icon.mute      | muted state of the volume control                                         |
| .gg-ez-vp--button-icon.low       | low volume state of the volume control (under 33%)                        |
| .gg-ez-vp--button-icon.medium    | medium volume state of the volume control (between 34% and 66%)           |
| .gg-ez-vp--button-icon.high      | high volume state of the volume control (above 66%)                       |
| .gg-ez-vp--button-icon.play      | shown while video is not playing on the play control                      |
| .gg-ez-vp--button-icon.pause     | shown while video is playing on the play control                          |
| .gg-ez-vp--button-icon.replay    | shown after the video ends                                                |
| .gg-ez-vp--button-icon.expand    | expand button icon                                                        |
| .gg-ez-vp--blocker               | invisible div that can be used to prevent clicks on VPAID/VAST thumbnails |
| .gg-ez-vp--slot                  | invisible div used as VPAID slot                                          |

There are also a few modifier selectors applied to the container `.gg-ez-vp`:

| selector               | description                                                                                                       |
| ---------------------- | ----------------------------------------------------------------------------------------------------------------- |
| .gg-ez-vp--no-scrub    | applied on `adControls: true`. Prevents displaying the scrub and adjusts ad styles.                               |
| .gg-ez-vp--skip        | applied when the skip button will be displayed                                                                    |
| .gg-ez-vp--volume-only | adjusts the styles when only the volume toggle button is displayed                                                |
| .gg-ez-vp--touchscreen | added when the device uses a touchscreen, this will show all enabled controls instead of activating them on hover |

See src/styles.css for all styles.

## Public methods

| method name      | parameters              | description                                                                                                                                                |
| ---------------- | ----------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------- |
| destroy          | none                    | removes all instance and video tag listeners, removes the player from the DOM leaving behind the original container                                        |
| fullscreenToggle | none                    | toggles the fullscreen mode                                                                                                                                |
| getCurrentTime   | none                    | returns the currentTime from the video tag, this data is also provided by the playback-progress event                                                      |
| muteUnmute       | none                    | toggle video sound on/off                                                                                                                                  |
| mute             | none                    | disable video sound (sets volume to 0)                                                                                                                     |
| unmute           | none                    | enable video sound (sets volume to last value)                                                                                                             |
| on               | `eventName, listenerFn` | attaches a listener function to either the video tag or the player instance, the function will be run every time the event is fired, [see events](#events) |
| once             | `eventName, listenerFn` | attaches a listener function to either the video tag or the player instance, the function will fire just on time, [see events](#events)                    |
| pause            | none                    | stop video playback                                                                                                                                        |
| playPause        | none                    | toggle pause state on/off                                                                                                                                  |
| play             | none                    | start video playback                                                                                                                                       |
| volume           | float number            | sets the video volume, [see volume configuration](#customization)                                                                                          |
| getVolume        | float number            | gets the video volume                                                                                                                                      |
| getDuration      | number                  | gets the video duration                                                                                                                                    |

## Accessible properties

All properties are read-only:

| property name | type     | description                                                                                                                                                    |
| ------------- | -------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| ready         | boolean  | helps identify when the video/ad is rendered and ready for playback                                                                                            |
| dataReady     | boolean  | helps identify when the player has all the data it needs to render the video and all listeners have been set                                                   |
| config        | object   | current player's configuration, see Options                                                                                                                    |
| player        | DOM node | current video tag, it is not recommended to interact directly with it, and instead rely on the player's methods, but is provided if necessary                  |
| dimensions    | object   | current player dimensions                                                                                                                                      |
| VPAIDWrapper  | object   | wrapper that allows direct interaction with a VPAID creative if available, it is not recommended to interact directly with it, but it is possible if necessary |

## Events

Listen for any `<video>` tag [events](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video#Events) or GgEzVp events:

```
ggEzVpInstance.on('play', myPlayListener);
ggEzVpInstance.on('playback-progress', myPlaybackListener);
```

Listen to them only once:

```
ggEzVpInstance.once('play', myPlayListener);
ggEzVpInstance.once('playback-progress', myProgressListener);
```

The player also emits custom events to extend the video tag behavior:

| event name        | description                                                                                                 | payload                                                                                   |
| ----------------- | ----------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------- |
| data-ready        | emitted when a source for the video is received, either by configuration or after asynchronous VAST parsing | `undefined`                                                                               |
| playback-progress | emitted when the video currentTime changes                                                                  | `{ remainingTime, readableTime, duration, currentTime, fancyCurrentTime, fancyDuration }` |
| player-click      | emitted when clicks are detected inside the container element                                               | click event                                                                               |
| pre-destroy       | emitted before removing listeners and the container node                                                    | `undefined`                                                                               |
| ready             | emitted when the class is ready for playback                                                                | `undefined`                                                                               |
| resize            | emitted when video tag changes either width or height                                                       | `{ width, height }`                                                                       |
| skip              | emitted when the skip button is pressed                                                                     | `undefined`                                                                               |
| expand            | emitted when the fullscreen is toggled                                                                      | boolean `isFullscreen`                                                                    |
| error             | emitted when player encounters an error                                                                     | error message or object                                                                   |


CHANGELOG.md
# Gg-Ez-Vp Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## [1.10.0] - 2020-12-17

-   Update VAST-client to 3.1.1 (#92)
-   Updates github actions because of a [vulnerability](https://github.blog/changelog/2020-10-01-github-actions-deprecating-set-env-and-add-path-commands/).

## [1.9.0] - 2020-08-24

-   Update build process and support IE 10/11 (#91)

## [1.8.0] - 2020-05-26

-   Add nonlinear creative support (#87)

## [1.7.6] - 2020-04-30

-   Add overlays for VAST (#86)

## [1.7.5] - 2020-04-16

-   Change VPAID volume after initAd call

## [1.7.4] - 2020-01-27

-   Emit playback-progress on VPAID milestones (#82)
-   Update VPAID event docs (#75)
-   Add timestamp separator if duration and currenTime exist

## [1.7.3] - 2020-01-24

-   invalidate CDN cache on releases

## [1.7.2] - 2020-01-22

-   fix VPAID clickthru when delegated to the player
-   hide timestamps when either duration or currentTime are not provided

## [1.7.1] - 2020-01-22

-   Listen for container resize and call VPAIDWrapper.resizeAd. Delegate listening for the container resize to https://marcj.github.io/css-element-queries/ (#76)

## [1.7.0] - 2020-01-10

-   Bugfix/vpaid rendering 3 (#72)

    -   track VAST errors

    -   prevent early calls to getAdDuration

    -   remove unused import

    -   use playerContainer container for VPAID creatives

    -   add vasttracker as a ggezvp property

    -   Use cached VPAID scripts

    -   Remove VAST load timestamp macros

    -   Guard against undefined VPAID wrapper method calls

    -   Attach VPAID listeners before attempting script load

    -   Reorder vpaid script listeners

    -   Guard against undefined VPAIDWrapper calls

    -   Don't call vpaid methods from instance before data-ready

    -   render VPAID in mobile safari

    -   Fix VPAID clickThrough in iOS, android chrome broken now

    -   Use independent div for VPAID creative slot

## [1.6.0] - 2019-12-26

-   Handle VPAID replay (#58) (#68)

## [1.5.0] - 2019-12-23

-   Prevent VAST load until container is visible (#60) (#67)

## [1.4.3] - 2019-12-20

-   Only last macro was replaced

## [1.4.2] - 2019-12-20

-   Link to demo page

## [1.4.1] - 2019-12-20

-   Change script validation to always use the last gg-ez-vp.js instead of throwing an error
-   Improve demo page with dynamic source edition
-   include and replace VAST [timestamp] macro (#60)

## [1.4.0] - 2019-12-18

-   Prevent context menu over video tag (#40)
-   Fix VAST/VPAID clickthru and tracking (#48)

## [1.3.0] - 2019-12-13

-   Set volume icons and mute toggle based on muted property (#56)
-   Remove MOAT references and process README with prettier (#54)
-   Bugfix/mobile autoplay electric boogaloo (#52)

## [1.2.1] - 2019-12-09

-   use production script in demo page
-   Prevent ad timestamp display when isNaN (#51)
-   Feature/optional ad controls (#50)
-   Bugfix/firefox fullscreen toggle (#39)

    -   push ad controls to the top #38

    -   toggle expand visibility on ad controls

    -   allow fullscreen when inside iframe

    -   improve moz fullscreen comments

## [1.1.5] - 2019-11-22

-   README URL fixes

## [1.1.2] - 2019-11-22

-   Create workflow to deploy develop to `/vp/beta/`

## [1.1.1] - 2019-11-21

-   Display enabled controls by default in touchscreen devices.
-   Center video in portrait + fullscreen mode

## [1.1.0] - 2019-11-20

### Breaking change

-   Preloads volume/play state icons by reading DOM script tags. This will break compatibility with ESM and CJS modules.
-   Create a changelog


src/lib/VPAIDWrapper.js
import checkVPAIDInterface from '../helpers/checkVPAIDInterface';
import secondsToReadableTime from '../helpers/secondsToReadableTime';
import setCallbacksForCreative from './setCallbacksForCreative';
import {
    DATA_READY,
    ERROR,
    PLAYBACK_PROGRESS,
    PLAYER_CLICK,
    VPAID_STARTED,
    SKIP,
    EXPAND
} from '../constants';

export default class VPAIDWrapper {
    constructor({
        VPAIDCreative,
        emitter,
        dimensions: { width: containerWidth, height: containerHeight },
        creativeVersion,
        VASTTracker
    }) {
        this._creative = VPAIDCreative;
        this.emitter = emitter;
        const isValidVPAID = this.__checkVPAIDInterface(VPAIDCreative);
        if (!isValidVPAID) {
            /* eslint-disable-next-line no-console */
            console.log("GgEzVp [WARN]: The VPAIDCreative doesn't conform to the VPAID spec");
            VASTTracker.errorWithCode(901);
            return;
        }
        this.VASTTracker = VASTTracker;
        this.__setCallbacksForCreative();
        this.emitter.on(SKIP, this.skipAd);
        this.emitter.on(EXPAND, isExpanded => {
            const expandAdArgs = isExpanded
                ? [window.screen.width, window.screen.height, 'fullscreen']
                : [containerWidth, containerHeight, 'thumbnail'];
            if (parseFloat(creativeVersion) < 2) {
                return this.stopAd();
            }
            this.expandAd(...expandAdArgs);
        });
        // listen for window closing to send close events
        window.onbeforeunload = () => {
            VASTTracker.close();
        };
    }

    __checkVPAIDInterface = checkVPAIDInterface;

    __setCallbacksForCreative = setCallbacksForCreative;

    initAd = (width, height, viewMode, desiredBitrate, creativeData, environmentVars) => {
        this._creative?.initAd(
            width,
            height,
            viewMode,
            desiredBitrate,
            creativeData,
            environmentVars
        );
        this.setAdVolume(0);
    };

    /* Wrapper methods */

    // Pass through for startAd
    startAd() {
        this.emitter.emit(VPAID_STARTED);
        this._creative?.startAd();
    }

    // Pass through for stopAd
    stopAd() {
        this._creative?.stopAd();
    }

    // Pass through for getAdLinear
    getAdLinear() {
        return this._creative?.getAdLinear();
    }

    // Pass through for getAdRemainingTime
    getAdRemainingTime() {
        return this._creative?.getAdRemainingTime();
    }

    // Pass through for getAdExpanded
    getAdExpanded() {
        return this._creative?.getAdExpanded();
    }

    // Pass through for getAdSkippableState
    getAdSkippableState() {
        return this._creative?.getAdSkippableState();
    }

    skipAd() {
        if (this.getAdSkippableState()) {
            return this._creative?.skipAd();
        }
    }

    // Pass through for setAdVolume
    setAdVolume(val) {
        this._creative?.setAdVolume(val);
    }

    // Pass through for getAdVolume
    getAdVolume() {
        return this._creative?.getAdVolume();
    }

    // Pass through for getAdDuration
    getAdDuration() {
        return this._creative?.getAdDuration();
    }

    // Pass through for resizeAd
    resizeAd(width, height, viewMode) {
        this._creative?.resizeAd(width, height, viewMode);
    }

    // Pass through for pauseAd
    pauseAd() {
        this._creative?.pauseAd();
    }

    // Pass through for pauseAd
    resumeAd() {
        this._creative?.resumeAd();
    }

    // Pass through for expandAd
    expandAd() {
        this._creative?.expandAd();
    }

    // Pass through for collapseAd
    collapseAd() {
        this._creative?.collapseAd();
    }

    /* VPAID listeners */

    // Callback for AdPaused
    onAdPaused() {
        this.emitter.emit('AdPaused');
        this.emitter.emit('pause');
        this.VASTTracker.setPaused(true);
    }

    // Callback for AdPlaying
    onAdPlaying() {
        this.emitter.emit('AdPlaying');
        this.emitter.emit('play');
        this.VASTTracker.setPaused(false);
    }

    // Callback for AdError
    onAdError(message) {
        this.emitter.emit('AdError', message);
        this.emitter.emit(ERROR, message);
        this.VASTTracker.errorWithCode(901);
    }

    // Callback for AdLog
    onAdLog(message) {
        this.emitter.emit('AdLog', message);
    }

    // Callback for AdUserAcceptInvitation
    onAdUserAcceptInvitation() {
        this.emitter.emit('AdUserAcceptInvitation');
    }

    // Callback for AdUserMinimize
    onAdUserMinimize() {
        this.emitter.emit('AdUserMinimize');
    }

    // Callback for AdUserClose
    onAdUserClose() {
        this.emitter.emit('AdUserClose');
        this.VASTTracker.close();
    }

    // Callback for AdSkippableStateChange
    onAdSkippableStateChange() {
        const adSkippableState = this._creative?.getAdSkippableState();
        this.emitter.emit('AdSkippableStateChange', adSkippableState);
    }

    // Callback for AdExpandedChange
    onAdExpandedChange() {
        const adExpanded = this._creative?.getAdExpanded();
        this.emitter.emit('AdExpandedChange', adExpanded);
        this.VASTTracker.setFullscreen(adExpanded);
    }

    // Callback for AdSizeChange
    onAdSizeChange() {
        const width = this._creative?.getAdWidth();
        const height = this._creative?.getAdHeight();
        this.emitter.emit('AdSizeChange', { width, height });
    }

    // Callback for AdDurationChange
    onAdDurationChange() {
        const duration = this._creative?.getAdDuration();
        this.duration = duration;
        this.emitter.emit('AdDurationChange', duration);
        this.checkProgress();
        this.VASTTracker.setDuration(duration);
    }

    // Callback for AdRemainingTimeChange
    // the AdRemainingTimeChange event is NOT re-emitted, USE PLAYBACK_PROGRESS instead
    onAdRemainingTimeChange() {
        this.checkProgress();
    }

    // Checks the currentTime, sets it on the VASTTracker/VPAIDWrapper, then
    // emits: PLAYBACK_PROGRESS
    // { remainingTime, readableTime, duration, currentTime }
    // Should be called on AdRemainingTimeChange, AdDurationChange, 25%, 50% and 75%
    checkProgress = () => {
        const remainingTime = this._creative?.getAdRemainingTime();
        const duration = this._creative?.getAdDuration();
        const currentTime = duration - remainingTime;
        const fancyDuration = secondsToReadableTime(duration);
        const fancyCurrentTime = secondsToReadableTime(currentTime);
        const payload = {
            remainingTime,
            fancyCurrentTime,
            fancyDuration,
            duration,
            currentTime
        };
        this.currentTime = currentTime;
        this.emitter.emit(PLAYBACK_PROGRESS, payload);
        this.VASTTracker.setProgress(currentTime);
    };

    // Callback for AdImpression
    onAdImpression() {
        this.emitter.emit('AdImpression');
        this.VASTTracker.trackImpression();
    }

    // Callback for AdClickThru
    onAdClickThru(url, id, playerHandles) {
        this.VASTTracker.once('clickthrough', VASTClickURL => {
            // click priority:
            // 1- VPAID
            // 2- VAST
            // 3- fallback passed to VASTTracker.click(url)
            const clickUrl = url || VASTClickURL;
            const payload = { url: clickUrl, id, playerHandles };
            this.emitter.emit(PLAYER_CLICK, payload);
            if (playerHandles) {
                window.open(clickUrl);
            }
        });
        // url is used as a fallback if not provided by the VAST tag
        this.VASTTracker.click(url);
    }

    // Callback for AdInteraction
    onAdInteraction(id) {
        this.emitter.emit('AdInteraction', id);
    }

    // Callback for AdVideoStart
    onAdVideoStart() {
        // Video 0% completed
        this.emitter.emit('play');
        this.emitter.emit('AdVideoStart');
    }

    // Callback for AdVideoFirstQuartile
    onAdVideoFirstQuartile() {
        // Video 25% completed
        this.emitter.emit('AdVideoFirstQuartile');
        this.checkProgress();
    }

    // Callback for AdVideoMidpoint
    onAdVideoMidpoint() {
        // Video 50% completed
        this.emitter.emit('AdVideoMidpoint');
        this.checkProgress();
    }

    // Callback for AdVideoThirdQuartile
    onAdVideoThirdQuartile() {
        // Video 75% completed
        this.emitter.emit('AdVideoThirdQuartile');
        this.checkProgress();
    }

    // Callback for AdVideoComplete
    onAdVideoComplete() {
        // Video 100% completed
        this.emitter.emit('AdVideoComplete');
        this.checkProgress();
        this.VASTTracker.complete();
    }

    // Callback for AdLinearChange
    onAdLinearChange() {
        const adLinear = this._creative?.getAdLinear();
        this.emitter.emit('AdLinearChange', adLinear);
    }

    // Callback for AdLoaded
    onAdLoaded() {
        this.emitter.emit('AdLoaded');
        this.emitter.emit(DATA_READY);
    }

    // Callback for startAd
    onStartAd() {
        this.emitter.emit('AdStarted');
    }

    // Callback for stopAd
    onStopAd() {
        this.emitter.emit('ended');
        this.emitter.emit('AdStopped');
    }

    // Callback for skipAd
    onSkipAd() {
        this.emitter.emit('AdSkipped');
        this.VASTTracker.skip();
    }

    // Callback for AdVolumeChange
    onAdVolumeChange() {
        const volume = this._creative?.getAdVolume();
        const isMuted = volume == 0;
        this.emitter.emit('AdVolumeChange', volume);
        this.VASTTracker.setMuted(isMuted);
    }
}
